DatExplorer/Render/TextureCache.cs

```diff
@@ -527,35 +537,67 @@ public static Texture2D Get(uint fileID)
                     return swatch;
                 }
 
-                var surfaceTexture = DatManager.PortalDat.ReadFromDat<SurfaceTexture>(surface.OrigTextureId);
+                uint textureId = surface.OrigTextureId;
+                if (textureMapChanges != null)
+                {
+                    var tmChange = textureMapChanges.Where(t => t.OldTexture == surface.OrigTextureId).FirstOrDefault();
+                    if (tmChange != null)
+                        textureId = tmChange.NewTexture;
+                }
+                var surfaceTexture = DatManager.PortalDat.ReadFromDat<SurfaceTexture>(textureId);
 
-                return GetTexture(surface, surfaceTexture.Textures[0]);
+                return GetTexture(surface, surfaceTexture.Textures[0], customPaletteColors, useCache);
             }
             return null;
         }
 
-        private static Texture2D GetTexture(uint textureID)
+        /// <summary>
+        /// Helper function to determine if we use the "clean" surface from the portal.dat, or the customized one with a texture swap.
+        /// </summary>
+        /// <param name="surfaceID"></param>
+        /// <param name="surfaces"></param>
+        /// <returns></returns>
+        private static FileTypes.Surface GetCorrectSurface(uint surfaceID, List<FileTypes.Surface> surfaces = null)
+        {
+            var datSurface = DatManager.PortalDat.ReadFromDat<Surface>(surfaceID);
+            FileTypes.Surface surface = new FileTypes.Surface(datSurface);
+
+            if (surfaces != null && surfaces.Count > 0)
+            {
+                foreach (var s in surfaces)
+                {
+                    if (s._surface.Id == surfaceID)
+                        surface = s;
+                }
+            }
+
+            return surface;
+        }
+
+        private static Texture2D GetTexture(uint textureID, Dictionary<int, uint> customPaletteColors = null, bool useCache = true)
         {
-            if (Textures.TryGetValue(textureID, out var cached))
+            if (useCache && Textures.TryGetValue(textureID, out var cached))
                 return cached;
 
-            var texture = LoadTexture(textureID);
-            Textures.Add(textureID, texture);
+            var texture = LoadTexture(textureID, useCache);
```

Comment:
For this particular call, it should just be the original LoadTexture(textureID);

For this particular overload of LoadTexture() the 2nd param is bool useDummy, instead of bool useCache

useDummy=true is causing all of the textures to be invisible/flat textures in texture viewer, and for the world viewer terrain